Doc/howto/annotations.rst

```diff
@@ -0,0 +1,205 @@
+  for example ``getattr(o, '__annotations__', None)``.
+
+
+Accessing The Annotations Dict Of An Object In Python 3.9 And Older
+===================================================================
+
+  In Python 3.9 and older, accessing the annotations dict
+  of an object is much more complicated than in newer versions.
+  The problem is a design flaw in these older versions of Python,
+  specifically to do with class annotations.
+
+  Best practice for accessing the annotations dict of other
+  objects--functions, other callables, and modules--is the same
+  as best practice for 3.10, assuming you aren't calling
+  :func:`inspect.get_annotations`: you should use three-argument
+  :func:`getattr` to access the object's ``__annotations__``
+  attribute.
+
+  Unfortunately, this isn't best practice for classes.  The problem
+  is that, since ``__annotations__`` is optional on classes, and
+  because classes can inherit attributes from their base classes,
+  accessing the ``__annotations__`` attribute of a class may
+  inadvertently return the annotations dict of a *base class.*
+  As an example::
+
+      class Base:
+          a:int = 3
+          b:str = 'abc'
+
+      class Derived(Base):
+          pass
+
+      print(Derived.__annotations__)
+
+  This will print the annotations dict from ``Base``, not
+  ``Derived``.
+
+  Your code will have to have a separate code path if the object
+  you're examining is a class (``isinstance(o, type)``).
+  In that case, best practice relies on an implementation detail
+  of Python 3.9 and before: if a class has annotations defined,
+  they are stored in the class's ``__dict__`` dictionary.  Since
+  the class may or may not have annotations defined, best practice
+  is to call the ``get`` method on the class dict.
+
+  To put it all together, here is some sample code that safely
+  accesses the ``__annotations__`` attribute on an arbitrary
+  object in Python 3.9 and before::
+
+      if isinstance(o, type):
+          ann = o.__dict__.get('__annotations__', None)
+      else:
+          ann = getattr(o, '__annotations__', None)
+
+  After running this code, ``ann`` should be either a
+  dictionary or ``None``.  You're encouraged to double-check
+  the type of ``ann`` using :func:`isinstance` before further
+  examination.
+
+  Note that some exotic or malformed type objects may not have
+  a ``__dict__`` attribute, so for extra safety you may also wish
+  to use :func:`getattr` to access ``__dict__``.
+
+
+Manually Un-Stringizing Stringized Annotations
+==============================================
+
+  In situations where some annotations may be "stringized",
+  and you wish to evaluate those strings to produce the
+  Python values they represent, it really is best to
+  call :func:`inspect.get_annotations` to do the evaluations
+  for you.  Un-stringizing is a complicated process, and
+  it's best to let :func:`inspect.get_annotations` do it
+  for you.
```

Comment:
```suggestion
  for you.
```
This sentence is just repetitive

Reply:
I'm also not sure I agree with the sentiment. There's a few plausible scenarios where it won't work:
- The annotation uses a name defined in a nested scope that's no longer accessible (one of the scenarios that PEP 649 would fix)
- The annotation uses a name defined in `if TYPE_CHECKING`
- The annotation uses PEP 604 syntax `int | str` and `from __future__ import annotations` before Python 3.10
In such cases, eval will throw an error at runtime, but if you try hard enough you can still extract useful information. In my own tool (pyanalyze) I instead `ast.parse` string annotations and evaluate them manually.

I would recommend at least adding a warning that `eval()` may throw for some annotations that static type checkers will process correctly.

Reply:
I followed all your suggestions, except adding references from datamodel.rst.  Again, I'm not sure how to work it in in a way that reads well.  Can you give a concrete example of how you'd add a reference to this HOWTO?

p.s. I also found one or two spots where I added a contraction.  When I get exacting, I tend to stop using contractions, and which I think makes my text sound stiff.

p.p.s. Did I forget any quirks?

Reply:
> Can you give a concrete example of how you'd add a reference to this HOWTO?

How about adding "For more details on how to use this attribute, see <link>" to the cell for `__annotations__` in the table under https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy ?